File: spring-data-requery-java/src/test/java/com/coupang/springframework/data/requery/domain/upsert/UpsertTest.java
package com.coupang.springframework.data.requery.domain.upsert;

import com.coupang.springframework.data.requery.domain.AbstractDomainTest;
import org.junit.Before;
import org.junit.Test;

import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;


/**
 * @author Diego on 2018. 6. 10..
 */
public class UpsertTest extends AbstractDomainTest {

    @Before
    public void setup() {
        requeryTemplate.deleteAll(UpsertTag.class);
        requeryTemplate.deleteAll(UpsertEvent.class);
        requeryTemplate.deleteAll(UpsertPlace.class);
        requeryTemplate.deleteAll(UpsertLocation.class);
    }

    @Test
    public void insert_location_with_embedded_address() {
        UpsertLocation location = new UpsertLocation();
        location.setName("Tower 730");
        location.getAddress().setZipcode("12345");
        location.getAddress().setCity("seoul");

        requeryTemplate.insert(location);

        assertThat(location.isNew()).isFalse();

        UpsertLocation loaded = requeryTemplate.findById(UpsertLocation.class, location.getId());
        assertThat(loaded).isEqualTo(location);
        assertThat(loaded.getAddress().getZipcode()).isEqualTo("12345");
    }

    @Test
    public void insert_many_to_many() {
        UUID eventId = UUID.randomUUID();
        UpsertEvent event = new UpsertEvent();
        event.setId(eventId);
        event.setName("test");

        UpsertTag tag1 = new UpsertTag();
        tag1.setId(UUID.randomUUID());
        tag1.setName("tag1");
        UpsertTag tag2 = new UpsertTag();
        tag2.setId(UUID.randomUUID());
        tag2.setName("tag2");

        event.getTags().add(tag1);
        event.getTags().add(tag2);

        requeryTemplate.insert(event);

        assertThat(requeryTemplate.count(UpsertEvent.class).get().value()).isEqualTo(1);
        assertThat(requeryTemplate.count(UpsertTag.class).get().value()).isEqualTo(2);

        UpsertEvent loaded = requeryTemplate.findById(UpsertEvent.class, event.getId());

        assertThat(loaded.getTags()).hasSize(2).containsOnly(tag1, tag2);

        requeryTemplate.delete(loaded);
        requeryTemplate.refreshAll(loaded);

        assertThat(requeryTemplate.count(UpsertEvent.class).get().value()).isEqualTo(0);
        assertThat(requeryTemplate.count(UpsertTag.class).get().value()).isEqualTo(0);
    }

    @Test
    public void upsert_event() {
        UUID eventId = UUID.randomUUID();
        UpsertEvent event = new UpsertEvent();
        event.setId(eventId);
        event.setName("test");

        requeryTemplate.upsert(event);

        assertThat(requeryTemplate.count(UpsertEvent.class).get().value()).isEqualTo(1);

        UpsertEvent loaded = requeryTemplate.findById(UpsertEvent.class, event.getId());
        assertThat(loaded).isEqualTo(event);

        requeryTemplate.delete(loaded);
        assertThat(requeryTemplate.count(UpsertEvent.class).get().value()).isEqualTo(0);
    }

    @Test
    public void upsert_one_to_many() {
        UUID eventId = UUID.randomUUID();
        UpsertEvent event = new UpsertEvent();
        event.setId(eventId);
        event.setName("test");

        UpsertPlace place = new UpsertPlace();
        place.setId(UUID.randomUUID().toString());
        place.setName("place");

        place.getEvents().add(event);

        requeryTemplate.upsert(place);
        UpsertPlace savedPlace = requeryTemplate.findById(UpsertPlace.class, place.getId());

        assertThat(savedPlace.getId()).isEqualTo(place.getId());
        assertThat(savedPlace.getEvents()).hasSize(1);
        assertThat(savedPlace.getEvents().iterator().next().getId()).isEqualTo(eventId);
    }

    @Test
    public void upsert_many_to_many() {
        UUID eventId = UUID.randomUUID();
        UpsertEvent event = new UpsertEvent();
        event.setId(eventId);
        event.setName("test");

        UpsertTag tag1 = new UpsertTag();
        tag1.setId(UUID.randomUUID());
        tag1.setName("tag1");
        UpsertTag tag2 = new UpsertTag();
        tag2.setId(UUID.randomUUID());
        tag2.setName("tag2");

        event.getTags().add(tag1);
        event.getTags().add(tag2);

        requeryTemplate.upsert(event);
        assertThat(requeryTemplate.count(UpsertEvent.class).get().value()).isEqualTo(1);
        assertThat(requeryTemplate.count(UpsertTag.class).get().value()).isEqualTo(2);

        UpsertEvent loaded = requeryTemplate.findById(UpsertEvent.class, event.getId());

        requeryTemplate.delete(loaded);

        assertThat(requeryTemplate.count(UpsertEvent.class).get().value()).isEqualTo(0);
        assertThat(requeryTemplate.count(UpsertTag.class).get().value()).isEqualTo(0);
    }

    @Test
    public void upsert_insert_one_to_many() {
        UUID eventId = UUID.randomUUID();
        UpsertEvent event = new UpsertEvent();
        event.setId(eventId);
        event.setName("test");

        UpsertPlace place = new UpsertPlace();
        place.setId(UUID.randomUUID().toString());
        place.setName("place");

        place.getEvents().add(event);

        requeryTemplate.insert(place);

        UpsertPlace savedPlace = requeryTemplate.findById(UpsertPlace.class, place.getId());

        assertThat(savedPlace.getId()).isEqualTo(place.getId());
        assertThat(savedPlace.getEvents()).hasSize(1);
        assertThat(savedPlace.getEvents().iterator().next().getId()).isEqualTo(eventId);
    }

    @Test
    public void upsert_one_to_many_with_empty_collection() {
        UUID eventId = UUID.randomUUID();
        UpsertEvent event = new UpsertEvent();
        event.setId(eventId);
        event.setName("test");

        UpsertPlace place = new UpsertPlace();
        place.setId(UUID.randomUUID().toString());

        place.getEvents().add(event);
        place.getEvents().remove(event);

        requeryTemplate.upsert(place);
        UpsertPlace savedPlace = requeryTemplate.findById(UpsertPlace.class, place.getId());

        assertThat(savedPlace.getId()).isEqualTo(place.getId());
        assertThat(savedPlace.getEvents()).hasSize(0);
    }

    @Test
    public void upsert_exists_entity() {
        UUID eventId = UUID.randomUUID();
        UpsertEvent event = new UpsertEvent();
        event.setId(eventId);
        event.setName("event1");

        requeryTemplate.insert(event);

        UpsertEvent event2 = new UpsertEvent();
        event2.setId(eventId);
        event2.setName("event2");

        requeryTemplate.upsert(event2);

        Iterable<UpsertEvent> events = requeryTemplate.findAll(UpsertEvent.class);
        assertThat(events).hasSize(1).containsOnly(event2);
    }

}

File: spring-data-requery-kotlin/src/test/kotlin/com/coupang/springframework/data/requery/domain/upsert/UpsertTest.kt
package com.coupang.springframework.data.requery.domain.upsert

import com.coupang.springframework.data.requery.domain.AbstractDomainTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import java.util.*

/**
 * UpsertTest
 *
 * @author debop@coupang.com
 * @since 18. 5. 31
 */
class UpsertTest: AbstractDomainTest() {

    @Before
    fun setup() {
        with(requeryKotlin) {
            deleteAll(UpsertTag::class)
            deleteAll(UpsertEvent::class)
            deleteAll(UpsertPlace::class)
            deleteAll(UpsertLocation::class)
        }
    }

    @Test
    fun `insert location with embedded address`() {
        val location = UpsertLocation().apply {
            name = "Tower 730"
            address.zipcode = "12345"
            address.city = "seoul"
        }

        with(requeryKotlin) {
            insert(location)

            assertThat(location.isNew()).isFalse()

            val loaded = findById(UpsertLocation::class, location.id)!!
            assertThat(loaded).isEqualTo(location)
            assertThat(loaded.address.zipcode).isEqualTo("12345")
        }
    }

    @Test
    fun `insert many to many`() {
        val eventId = UUID.randomUUID()
        val event = UpsertEvent().apply {
            id = eventId
            name = "test"
        }

        val tag1 = UpsertTag().apply { id = UUID.randomUUID(); name = "tag1" }
        val tag2 = UpsertTag().apply { id = UUID.randomUUID(); name = "tag2" }

        event.tags.add(tag1)
        event.tags.add(tag2)

        with(requeryKotlin) {

            insert(event)

            assertThat(count(UpsertEvent::class).get().value()).isEqualTo(1)
            assertThat(count(UpsertTag::class).get().value()).isEqualTo(2)

            val loaded = findById(UpsertEvent::class, event.id)!!

            assertThat(loaded.tags).hasSize(2).containsOnly(tag1, tag2)

            delete(loaded)
            refreshAll(loaded)

            assertThat(count(UpsertEvent::class).get().value()).isEqualTo(0)
            assertThat(count(UpsertTag::class).get().value()).isEqualTo(0)
        }
    }

    @Test
    fun `upsert event`() {
        val eventId = UUID.randomUUID()
        val event = UpsertEvent().apply {
            id = eventId
            name = "test"
        }
        with(requeryKotlin) {
            upsert(event)

            assertThat(count(UpsertEvent::class).get().value()).isEqualTo(1)

            val loaded = findById(UpsertEvent::class, event.id)!!
            assertThat(loaded).isEqualTo(event)

            delete(loaded)
            assertThat(count(UpsertEvent::class).get().value()).isEqualTo(0)
        }
    }

    @Test
    fun `upsert one to many`() {
        val eventId = UUID.randomUUID()
        val event = UpsertEvent().apply {
            id = eventId
            name = "test"
        }

        val place = UpsertPlace().apply {
            id = UUID.randomUUID().toString()
            name = "place"
        }

        place.events.add(event)

        with(requeryKotlin) {

            upsert(place)
            val savedPlace = findById(UpsertPlace::class, place.id)!!

            assertThat(savedPlace.id).isEqualTo(place.id)
            assertThat(savedPlace.events).hasSize(1)
            assertThat(savedPlace.events.firstOrNull()?.id).isEqualTo(eventId)
        }
    }

    @Test
    fun `upsert many to many`() {
        val eventId = UUID.randomUUID()
        val event = UpsertEvent().apply {
            id = eventId
            name = "test"
        }

        val tag1 = UpsertTag().apply { id = UUID.randomUUID(); name = "tag1" }
        val tag2 = UpsertTag().apply { id = UUID.randomUUID(); name = "tag2" }

        event.tags.add(tag1)
        event.tags.add(tag2)

        with(requeryKotlin) {

            upsert(event)

            assertThat(count(UpsertEvent::class).get().value()).isEqualTo(1)
            assertThat(count(UpsertTag::class).get().value()).isEqualTo(2)

            val loaded = findById(UpsertEvent::class, event.id)!!

            assertThat(loaded.tags).hasSize(2).containsOnly(tag1, tag2)

            delete(loaded)

            assertThat(count(UpsertEvent::class).get().value()).isEqualTo(0)
            assertThat(count(UpsertTag::class).get().value()).isEqualTo(0)
        }
    }

    @Test
    fun `upsert insert one to many`() {

        val eventId = UUID.randomUUID()
        val event = UpsertEvent().apply {
            id = eventId
            name = "test"
        }

        val place = UpsertPlace().apply {
            id = UUID.randomUUID().toString()
            name = "place"
        }

        place.events.add(event)

        with(requeryKotlin) {

            insert(place)


            val savedPlace = findById(UpsertPlace::class, place.id)!!

            assertThat(savedPlace.id).isEqualTo(place.id)
            assertThat(savedPlace.events).hasSize(1)
            assertThat(savedPlace.events.firstOrNull()?.id).isEqualTo(eventId)
        }
    }

    @Test
    fun `upsert one to many with empty collection`() {
        val eventId = UUID.randomUUID()
        val event = UpsertEvent().apply {
            id = eventId
            name = "test"
        }

        val place = UpsertPlace().apply {
            id = UUID.randomUUID().toString()
            name = "place"
        }
        place.events.add(event)
        place.events.remove(event)

        with(requeryKotlin) {

            upsert(place)
            val savedPlace = findById(UpsertPlace::class, place.id)!!

            assertThat(savedPlace.id).isEqualTo(place.id)
            assertThat(savedPlace.events).hasSize(0)
        }
    }

    @Test
    fun `upsert exists entity`() {
        val eventId = UUID.randomUUID()
        val event1 = UpsertEvent().apply {
            id = eventId
            name = "event1"
        }

        with(requeryKotlin) {
            insert(event1)

            val event2 = UpsertEvent().apply {
                id = eventId
                name = "event2"
            }

            upsert(event2)

            val events = findAll(UpsertEvent::class)
            assertThat(events).hasSize(1).containsOnly(event2)
        }
    }
}